### DISCUSSION TITLE
Error in Tutorial section "Make the Project Installable"

### DISCUSSION THREAD

**Discussion by prcantwell** [pallets/flask#5099] [Q&A]
While following the instructions for [Make the Project Installable](https://flask.palletsprojects.com/en/2.3.x/tutorial/install/) in the  [Flask Tutorial](https://flask.palletsprojects.com/en/2.3.x/tutorial/), I encountered the error below when running `pip install -e .`:

```
(env) PS D:\repos-tutorials\flask-tutorial> pip install -e .
Obtaining file:///D:/repos-tutorials/flask-tutorial
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      error: Multiple top-level packages discovered in a flat-layout: ['flaskr', 'instance'].
      
      To avoid accidental inclusion of unwanted files or directories,
      setuptools will not proceed with this build.
     
      If you are trying to create a single distribution with multiple packages
      on purpose, you should not rely on automatic discovery.
     
      1. set up custom discovery (`find` directive with `include` or `exclude`)
      2. use a `src-layout`
      3. explicitly set `py_modules` or `packages` with a list of names
     
      To find more information, look for "package discovery" on setuptools docs.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

The error message suggests three solutions.  Following the [setuptools docs](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html), I tried solution 3 by adding these two lines to the end of my `pyproject.toml` file:

```
[tool.setuptools]
packages = ['flaskr']
```

It worked (installation proceeded without errors).  

But is this the correct way to fix this error?  Or should I not have had this error in the first place?  (did I make a mistake following the tutorial?)

I'm using Python 3.9.2, setuptools 65.5.1, and pip 23.1.2 on Windows 10.

**Comment by alexqfredrickson** [comment]
`pip install -e .` seems to be really finnicky.  I've been encountering and working through a similar issue - which I resolved this by way of following the guidance at https://packaging.python.org/en/latest/tutorials/packaging-projects/ very carefully.  In my case, `pip install -e .` was requiring that my code was wrapped in a `src/your_projectname` directory at the project root, with `__init__.py` at each module root.  Might be worth trying in your case.
